enterprise/backend/src/metabase_enterprise/caching/strategies.clj

```diff
@@ -0,0 +1,114 @@
+                 [:min_duration ms/PositiveInt]]]
+     [:duration [:map
+                 [:type keyword?]
+                 [:duration ms/PositiveInt]
+                 [:unit [:enum "hours" "minutes" "seconds" "days"]]]]
+     [:schedule [:map
+                 [:type keyword?]
+                 [:last-expired-at [:maybe some?]]
+                 [:schedule u.cron/CronScheduleString]]]
+     [:query    [:map
+                 [:type keyword?]
+                 [:last-expired-at [:maybe some?]]
+                 [:field_id int?]
+                 [:aggregation [:enum "max" "count"]]
+                 [:schedule u.cron/CronScheduleString]]]]]))
+
+(defn granular-duration-hours
+  "Returns the granular cache duration (in hours) for a card. On EE, this first checking whether there is a stored value
+   for the card, dashboard, or database (in that order of decreasing preference). Returns nil on OSS."
+  [card dashboard-id]
+  (or (:cache_ttl card)
+      (:cache_ttl (t2/select-one [:model/Dashboard :cache_ttl] :id dashboard-id))
+      (:cache_ttl (t2/select-one [:model/Database :cache_ttl] :id (:database_id card)))))
+
+(defenterprise-schema granular-cache-strategy :- [:maybe CacheStrategy]
+  "Returns the granular cache strategy for a card."
+  :feature :cache-granular-controls
+  [card dashboard-id]
+  (when (public-settings/enable-query-caching)
+    (let [qs     (for [[i model model-id] [[1 "question"   (:id card)]
+                                           [2 "dashboard"  dashboard-id]
+                                           [3 "database"   (:database_id card)]
+                                           [4 "root"       0]]
+                       :when              model-id]
+                   {:from   [:cache_config]
+                    :select [:id
+                             [[:inline i] :ordering]]
+                    :where  [:and
+                             [:= :model model]
+                             [:= :model_id model-id]]})
+          q      {:select   [:id]
+                  :limit    1
+                  :from     [[{:union-all qs} :unused_alias]]
+                  :order-by :ordering}
+          config (t2/select-one :model/CacheConfig :id q)]
+      (if config
+        (merge {:type            (:strategy config)
+                :last-expired-at (:last_expired_at config)}
+               (:config config))
+        (when-let [ttl (granular-duration-hours card dashboard-id)]
+          {:type     :duration
+           :duration ttl
+           :unit     "hours"})))))
+
+;;; Strategy execution
+
+(defmulti fetch-cache-stmt-ee*
+  "Generate prepared statement for a db cache backend for a given strategy"
+  (fn [strategy _hash _conn] (:type strategy)))
+
+(defmethod fetch-cache-stmt-ee* :ttl [strategy query-hash conn]
+  (backend.db/fetch-cache-stmt-ttl strategy query-hash conn))
+
+(defmethod fetch-cache-stmt-ee* :duration [strategy query-hash conn]
+  (assert (:duration strategy))
```

Comment:
I am not sure about having these asserts. Does the QP guard error against these asserts? failed to fetch cache shouldn't make a query fail.

Reply:
Agree, it's now just a debug log in b3d0e32fcf